src/Servers/Kestrel/Core/src/Internal/Http3/Http3PendingStream.cs

```diff
@@ -0,0 +1,100 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Net.Http;
+using Microsoft.AspNetCore.Connections;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3;
+
+internal sealed class Http3PendingStream
+{
+    private readonly CancellationTokenSource abortedToken = new();
+
+    private ConnectionAbortedException? _abortedException;
+    private bool _isClosed;
+
+    internal readonly Http3StreamContext Context;
+    internal readonly long StreamId;
+    internal long StreamTimeoutTicks;
+
+    public Http3PendingStream(Http3StreamContext context, long id)
+    {
+        Context = context;
+        StreamTimeoutTicks = default;
+        StreamId = id;
+    }
+
+    public void Abort(ConnectionAbortedException exception)
+    {
+        if (_isClosed)
+        {
+            return;
+        }
+        _isClosed = true;
+
+        _abortedException = exception;
+
+        abortedToken.Cancel();
+
+        Context.Transport.Input.Complete(exception);
+        Context.Transport.Output.Complete(exception);
+    }
+
+    public async ValueTask<long> ReadNextStreamHeaderAsync(Http3StreamContext context, long streamId, Http3StreamType? advanceOn)
+    {
+        var Input = context.Transport.Input;
+        var advance = false;
+        SequencePosition consumed = default;
+        SequencePosition start = default;
+        try
+        {
+            while (!_isClosed)
+            {
+                var result = await Input.ReadAsync(abortedToken.Token);
```

Comment:
Is the token here required? Abort is completing the input and output with an exception. I think that will cause ReadAsync to exit.

Reply:
I tested it just now, apparently, the ReadAsync will not throw if you complete the PipeReader.

Reply:
Correction: Brennan suggested that I used the `PipeReader.CancelPendingRead()`. This does fix it and I can remove the TCS.